ledger/participant-integration-api/src/main/resources/db/migration/postgres/V39__multiple_submitters.sql

```diff
@@ -0,0 +1,12 @@
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+alter table participant_events
+    alter column submitter type varchar array using array[submitter];
+alter table participant_events
+    rename column submitter to submitters;
+
+alter table participant_command_completions
+    alter column submitting_party type varchar array using array[submitting_party];
```

Comment:
I did a quick local test, and the `submitters` part of the index isn't being used anymore, because arrays cannot be indexed by a btree index, so the index "silently" doesn't _fully_ work anymore (`completion_offset` and `application_id` are still looked up in the index for the query).

Reply:
Good question. For H2, I had to manually drop and re-create the index. [Postgres documentation](https://www.postgresql.org/docs/9.1/sql-altertable.html) says

> Indexes and simple table constraints involving the column will be automatically converted to use the new column type by reparsing the originally supplied expression.

so the index is kept. However, a quick google search also says that btree indices are of limited use for arrays. In particular, it looks like they can't help with `@>` or `&&` operations. I did not investigate the performance implications of this.

Reply:
@gerolf-da How did you test whether an index is used for `submitters`? The only index that contains `submitter` I see is the deduplication index:
```
CREATE UNIQUE INDEX idx_transactions_deduplication
  ON ledger_entries (submitter, command_id, application_id);
```

The deduplication query uses array equality, and that seems to work with the default btree index:
```
# CREATE TABLE test (foo varchar[], bar varchar);
# CREATE INDEX test_index_full_array on test ((foo));
# CREATE INDEX test_index_first_element on test ((foo[1]));
# CREATE INDEX test_index_primitive on test ((bar));
# EXPLAIN ANALYZE SELECT * FROM test WHERE bar = 'x';
                                                         QUERY PLAN                                                          
-----------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on test  (cost=4.18..12.64 rows=4 width=64) (actual time=0.004..0.004 rows=0 loops=1)
   Recheck Cond: ((bar)::text = 'x'::text)
   ->  Bitmap Index Scan on test_index_primitive  (cost=0.00..4.18 rows=4 width=0) (actual time=0.002..0.002 rows=0 loops=1)
         Index Cond: ((bar)::text = 'x'::text)
 Planning Time: 1.005 ms
 Execution Time: 0.022 ms
(6 rows)

# EXPLAIN ANALYZE SELECT * FROM test WHERE foo[1]='x';
                                                           QUERY PLAN                                                            
---------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on test  (cost=4.18..12.64 rows=4 width=64) (actual time=0.005..0.005 rows=0 loops=1)
   Recheck Cond: ((foo[1])::text = 'x'::text)
   ->  Bitmap Index Scan on test_index_first_element  (cost=0.00..4.18 rows=4 width=0) (actual time=0.002..0.002 rows=0 loops=1)
         Index Cond: ((foo[1])::text = 'x'::text)
 Planning Time: 0.075 ms
 Execution Time: 0.017 ms
(6 rows)

# EXPLAIN ANALYZE SELECT * FROM test WHERE foo=array['x']::varchar[];
                                                          QUERY PLAN                                                          
------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on test  (cost=4.18..12.64 rows=4 width=64) (actual time=0.007..0.007 rows=0 loops=1)
   Recheck Cond: (foo = '{x}'::character varying[])
   ->  Bitmap Index Scan on test_index_full_array  (cost=0.00..4.18 rows=4 width=0) (actual time=0.005..0.005 rows=0 loops=1)
         Index Cond: (foo = '{x}'::character varying[])
 Planning Time: 0.106 ms
 Execution Time: 0.119 ms
(6 rows)
```

Reply:
But [the query ](https://github.com/digital-asset/daml/pull/8035/files#diff-7b03107bfd312b0beea0d4a64b28409f967b58fd1e466dc84ca1f9a362798903R53-R55) doesn't use array equality but intersection (`&&`). As you stated, btree indices don't work with that. 
Given that completions normally are intended to be tailed by applications and it's also filtered by application id, I have a hunch that it's not a big problem, but I have no data to back this up.

Reply:
I see, I missed the fact that the `idx_transactions_deduplication` index would be used in a command completion query. So this PR will probably indeed make the command completion queries slower.